File: lib/processes/keys.ex
defmodule Keys do
	#this module is designed to make it impossible to extract the private key.
	#unfortunately, that means that if this file crashes, your private key gets deleted.
	#make sure you back up your private key.
  use GenServer
  @name __MODULE__
	def db_location do "/keys" end
  def start_link() do GenServer.start_link(__MODULE__, :ok, [name: @name]) end
  def init(:ok) do
		x = DB.get_raw(db_location)
		if x == "" do
			DB.put_function(db_location, fn() -> %{pub: nil, priv: nil}	end)
			{:ok, {nil, nil}}
		else
			{:ok, {x.pub, nil}}
		end
	end
	def store(pub, priv, brainwallet) do DB.put_function(db_location, fn() -> %{pub: pub, priv: Encryption.sym_enc(brainwallet, priv)}	end) end
  def handle_cast({:load, pub, priv, brainwallet}, x) do
		#store on harddrive. If you already have a pub/priv pair loaded, it will NOT replace it.
		x = DB.get_raw(db_location)
		if x.priv == nil do	store(pub, priv, brainwallet) end
		{:noreply, {pub, priv}}
	end
	def handle_cast({:unlock, brainwallet}, _) do
		x = DB.get_raw(db_location)
		priv = Encryption.sym_dec(brainwallet, x.priv)
		{:noreply, {x.pub, priv}}
	end
	def handle_cast(:lock, {pub, priv}) do {:noreply, {pub, nil}}	end
	def handle_cast({:change_password, current, new}, {pub, priv}) do
		x = DB.get_raw(db_location)
		msg = "abc"
		cond do 
			x == "" -> {:noreply, {pub, priv}}
			x.priv == nil ->
				{pub, priv} = CryptoSign.new_key
				store(pub, priv, new)
				{:noreply, {pub, priv}}
			true ->
				new_priv = Encryption.sym_dec(current, x.priv)
				msg = "abc"
				if CryptoSign.verify(msg, CryptoSign.sign(msg, new_priv), x.pub) do
					priv = new_priv
					pub = x.pub
					store(pub, priv, new)
				end
				{:noreply, {pub, priv}}
		end
	end
	def handle_call({:ss, pub}, _from, x) do
    {_, priv} = x
		out = CryptoSign.shared_secret(pub, priv)
		{:reply, out, x}
	end
  def handle_call({:raw_sign, m}, _from, x) do
    {_, priv} = x
		cond do
			not is_binary(m) -> sig = "not binary"
			priv == nil -> sig = "need passphrase to unlock"
			true -> sig = CryptoSign.sign(m, priv)
		end
    {:reply, sig, x}
	end
  def handle_call({:sign, m}, _from, x) do
    {pub, priv} = x
		cond do
			not is_map(m) -> sig = "bad tx"
			priv == nil -> sig = "need passphrase to unlock"
			true -> sig = CryptoSign.sign_tx(m, pub, priv)
		end
    {:reply, sig, x}
	end
	def handle_call(:status, _from, x) do
		{_, priv} = x
		y = DB.get_raw(db_location)
		cond do
			y.priv == nil -> out = :empty
			priv == nil -> out = :locked
			true -> out = :unlocked
		end
		{:reply, out, x}
	end
  def handle_call(:pubkey, _from, x) do {:reply, elem(x, 0), x} end
  def pubkey do GenServer.call(@name, :pubkey) end
  def sign(m) do GenServer.call(@name, {:sign, m}) end
  def raw_sign(m) do GenServer.call(@name, {:raw_sign, m}) end
  def load(pub, priv, brainwallet) do GenServer.cast(@name, {:load, pub, priv, brainwallet}) end
  def unlock(brainwallet) do GenServer.cast(@name, {:unlock, brainwallet}) end
  def lock do GenServer.cast(@name, :lock) end
	def status do GenServer.call(@name, :status) end
	def change_password(current, new) do GenServer.cast(@name, {:change_password, current, new}) end
  def new(brainwallet) do change_password("", brainwallet) end
	def shared_secret(pub) do GenServer.call(@name, {:ss, pub}) end
  #def master do load("BCmhaRq42NNQe6ZpRHIvDxHBThEE3LDBN68KUWXmCTKUZvMI8Ol1g9yvDVTvMsZbqHQZ5j8E7sKVCgZMJR7lQWc=", "pRxnT/38wyd6lSbwfCEVvchAL7m7AMYuZeQKrJW/RO0=") end
end

File: erlang/src/keys/keys.erl
%the hard drive stores {f, pubkey, encrypted(privkey), encrypted("sanity")).
%the ram stores either {pubkey, privkey} or {pubkey, ""} depending on if this node is locked.
-module(keys).
-behaviour(gen_server).
-export([start_link/0,code_change/3,handle_call/3,handle_cast/2,handle_info/2,init/1,terminate/2, pubkey/0,sign/1,sign/2,raw_sign/1,load/3,unlock/1,lock/0,status/0,change_password/2,new/1,shared_secret/1]).
-define(LOC(), "keys.db").
-define(SANE(), <<"sanity">>).
start_link() -> gen_server:start_link({local, ?MODULE}, ?MODULE, ok, []).
code_change(_OldVsn, State, _Extra) -> {ok, State}.
terminate(_, _) -> io:format("keys died"), ok.
-record(f, {pub = "", priv = "", sanity = ""}).
init(ok) -> 
    X = db:read(?LOC()),
    if
        X == "" -> 
            K = #f{},%we should probably make a new key instead.
            db:save(?LOC(),K);
        true -> K = #f{pub=X#f.pub}
    end,
    {ok, K}.
store(Pub, Priv, Brainwallet) -> 
    X = #f{pub=Pub, priv=encryption:bin_enc(Brainwallet, Priv), sanity=encryption:bin_enc(Brainwallet, ?SANE())},
    db:save(?LOC(), X),
    Y = db:read(?LOC()),
    Y = X.
handle_call({ss, Pub}, _From, R) ->
    {reply, sign:shared_secret(Pub, R#f.priv), R};
handle_call({raw_sign, _}, _From, R) when R#f.priv=="" ->
    {reply, "need to unlock passphrase", R};
handle_call({raw_sign, M}, _From, X) when not is_binary(M) ->
    {reply, "not binary", X};
handle_call({raw_sign, M}, _From, R) ->
    {reply, sign:sign(M, R#f.priv), R};
handle_call({sign, M, Accounts}, _From, R) -> 
    {reply, sign:sign_tx(M, R#f.pub, R#f.priv, Accounts), R};
handle_call(status, _From, R) ->
    Y = db:read(?LOC()),
    Out = if
              Y#f.priv == "" -> empty;
              R#f.priv == "" -> locked;
              true -> unlocked
          end,
    {reply, Out, R};
handle_call(pubkey, _From, R) -> {reply, R#f.pub, R}.
handle_cast({load, Pub, Priv, Brainwallet}, _R) ->
    store(Pub, Priv, Brainwallet),
    {noreply, #f{pub=Pub, priv=Priv}};
handle_cast({new, Brainwallet}, _R) ->
    {Pub, Priv} = sign:new_key(),
    store(Pub, Priv, Brainwallet),
    {noreply, #f{pub=Pub, priv=Priv}};
handle_cast({unlock, Brainwallet}, _) ->
    X = db:read(?LOC()),
    ?SANE() = encryption:bin_dec(Brainwallet, X#f.sanity),
    Priv = encryption:bin_dec(Brainwallet, X#f.priv),%err
    {noreply, #f{pub=X#f.pub, priv=Priv}};
handle_cast(lock, R) -> {noreply, #f{pub=R#f.pub}};
handle_cast({change_password, Current, New}, R) ->
    X = db:read(?LOC()),
    ?SANE() = encryption:bin_dec(Current, X#f.sanity),
    Priv = encryption:bin_dec(Current, X#f.priv),
    store(R#f.pub, Priv, New),
    {noreply, R};
handle_cast(_, X) -> {noreply, X}.
handle_info(_, X) -> {noreply, X}.
pubkey() -> gen_server:call(?MODULE, pubkey).
sign(M) -> gen_server:call(?MODULE, {sign, M, dict:new()}).
sign(M, Accounts) -> gen_server:call(?MODULE, {sign, M, Accounts}).
raw_sign(M) -> gen_server:call(?MODULE, {raw_sign, M}).
load(Pub, Priv, Brainwallet) -> gen_server:cast(?MODULE, {load, Pub, Priv, Brainwallet}).
unlock(Brainwallet) -> gen_server:cast(?MODULE, {unlock, Brainwallet}).
lock() -> gen_server:cast(?MODULE, lock).
status() -> gen_server:call(?MODULE, status).
change_password(Current, New) -> gen_server:cast(?MODULE, {change_password, Current, New}).
new(Brainwallet) -> gen_server:cast(?MODULE, {new, Brainwallet}).
shared_secret(Pub) -> gen_server:call(?MODULE, {ss, Pub}).
